Philips.CodeAnalysis.MaintainabilityAnalyzers/Readability/AvoidMultipleLambdasOnSingleLineAnalyzer.cs

```diff
@@ -28,74 +28,48 @@ public class AvoidMultipleLambdasOnSingleLineAnalyzer : DiagnosticAnalyzer
 
 		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
 
-		private readonly HashSet<string> visitedTypes = new();
-
 		public override void Initialize(AnalysisContext context)
 		{
 			context.EnableConcurrentExecution();
 			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
 			
 			context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.SimpleLambdaExpression);
 			context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.ParenthesizedLambdaExpression);
-			context.RegisterCompilationAction(ClearCache);
 		}
 
 		private void Analyze(SyntaxNodeAnalysisContext context)
 		{
 			SyntaxNode node = context.Node;
-			TypeDeclarationSyntax typeDeclaration =
-				node.Ancestors().OfType<TypeDeclarationSyntax>().FirstOrDefault();
+			MethodDeclarationSyntax methodDeclaration =
+				node.Ancestors().OfType<MethodDeclarationSyntax>().FirstOrDefault();
 			
-			// Prevent reporting the same class twice.
-			string typeName = $"{node.SyntaxTree.FilePath}-{typeDeclaration?.Identifier.Text}";
-			lock (visitedTypes)
-			{
-				if (visitedTypes.Contains(typeName))
-				{
-					return;
-				}
-
-				visitedTypes.Add(typeName);
-			}
-
-
-			var lambdas = typeDeclaration?.DescendantNodes().OfType<LambdaExpressionSyntax>();
+			var lambdas = methodDeclaration?.DescendantNodes().OfType<LambdaExpressionSyntax>();
 			if (lambdas == null || !lambdas.Any())
 			{
 				return;
 			}
 
-			foreach (LambdaExpressionSyntax violation in FindLambdasOnSameLine(lambdas))
+			if (FindLambdasOnSameLine(node, lambdas))
 			{
-				Location loc = violation.GetLocation();
+				Location loc = node.GetLocation();
 				context.ReportDiagnostic(Diagnostic.Create(Rule, loc));
 			}
 		}
 
-		private void ClearCache(CompilationAnalysisContext _)
-		{
-			lock (visitedTypes)
-			{
-				visitedTypes.Clear();
-			}
-		}
-
-		private static IEnumerable<LambdaExpressionSyntax> FindLambdasOnSameLine(IEnumerable<LambdaExpressionSyntax> lambdas)
+		private static bool FindLambdasOnSameLine(SyntaxNode ourLambda, IEnumerable<LambdaExpressionSyntax> lambdas)
 		{
-			// Using HashSet to filter out duplicates.
-			// And relying on the fact that the order in the SyntaxTree is the same as in the .cs file.
-			HashSet<LambdaExpressionSyntax> result = new();
 			int previousLine = -1;
 			foreach(LambdaExpressionSyntax lambda in lambdas)
 			{
+				// If we find a violation, report based on the heuristic that the violator is our lambda.
 				int currentLine = lambda.GetLocation().GetLineSpan().EndLinePosition.Line;
-				if (previousLine == currentLine)
+				if (previousLine == currentLine && lambda == ourLambda)
```

Comment:
I would expect lambda != ourLambda.

If it is NOT equal, it means there is another lambda on the same line.
The current implementation just return always true in a complicated way as you're moving to the parent and back down again and return true if you find yourself...